Data Dictionary of original dataset
| Variable | Description | Details |
| -------- | ----------- | ------- |
| Survival | Survival | 0 = No; 1 = Yes |
| Pclass | Passenger Class | 1 = upper; 2 = middle; 3 = lower |
| Name | First and Last Name | |
| Sex | Sex | |
| Age | Age | Fractional if Age less than One (1); If the Age is Estimated, it is in the form xx.5 |
| Sibsp | Number of Siblings/Spouses Aboard | |
| Parch | Number of Parents/Children Aboard | |
| Ticket | Ticket Number | |
| Fare | Passenger Fare | |
| Cabin | Cabin | |
| Embarked | Port of Embarkation | C = Cherbourg; Q = Queenstown; S = Southampton |


In [1]:
# For data manipulation and analysis
import pandas as pd 
# For mathematical operations
import numpy as np 

# For interactive plots
import plotly as py 
# For easy-to-use, high-level interface for creating expressive and interactive visualizations
import plotly.express as px 
# For creating complex and customized plots
import plotly.graph_objects as go 
# For creating subplots in a single figure
from plotly.subplots import make_subplots 

In [2]:
# Load Dataset
train = pd.read_csv("train.csv")

# Make a deep copy of the train DataFrame and assign it to df
# So that we can always go back to the original dataset if we need to
df = train.copy(deep=True)

In [3]:
# Create a new column named FamilySize in df
'''
Add the values of SibSp (number of siblings and spouses aboard) 
and Parch (number of parents and children aboard) columns for each row 
and add 1 to get the total family size
'''

df["FamilySize"] = df["SibSp"] + df["Parch"] + 1 

In [4]:
# Obtain basic statistics for numerical variables
df.describe()

PassengerId    Survived      Pclass         Age       SibSp  \
count   891.000000  891.000000  891.000000  714.000000  891.000000   
mean    446.000000    0.383838    2.308642   29.699118    0.523008   
std     257.353842    0.486592    0.836071   14.526497    1.102743   
min       1.000000    0.000000    1.000000    0.420000    0.000000   
25%     223.500000    0.000000    2.000000   20.125000    0.000000   
50%     446.000000    0.000000    3.000000   28.000000    0.000000   
75%     668.500000    1.000000    3.000000   38.000000    1.000000   
max     891.000000    1.000000    3.000000   80.000000    8.000000   

            Parch        Fare  FamilySize  
count  891.000000  891.000000  891.000000  
mean     0.381594   32.204208    1.904602  
std      0.806057   49.693429    1.613459  
min      0.000000    0.000000    1.000000  
25%      0.000000    7.910400    1.000000  
50%      0.000000   14.454200    1.000000  
75%      0.000000   31.000000    2.000000  
max      6.000000  512.329200   11.000000

---

Objective : Analyze the missing values in a dataframe<br>
About the code : Create a function that can analyze the missing values in a dataframe and display them in a table and a heatmap. <br> 
Justification : The heatmap was chosen because it can visually show the distribution and pattern of missing values across the dataframe. 

Steps to build the function :
- The first line defines a function called `missing_values` that takes a dataframe as an input parameter.
- The second line creates a new dataframe with the data types of each column in the input dataframe using the `pd.DataFrame` method and assigns it to the variable `missingInfo`.
- The third line adds a new column to the `missingInfo` dataframe with the number of missing values in each column of the input dataframe using the `df.isnull().sum()` method.
- The fourth line adds another new column to the `missingInfo` dataframe with the percentage of missing values in each column of the input dataframe using the `round((df.isnull().sum() / len(df)) * 100, 2)` expression.
- The fifth line sorts the `missingInfo` dataframe by the percentage of missing values in descending order using the `sort_values` method with the parameter `by="% missing"` and `ascending=False`.
- The sixth line creates a heatmap plot of the input dataframe showing which values are missing (True) or present (False) using the `px.imshow` method with the parameters `df.isnull()`, `template="plotly_dark"`, `width=500`, and `title="Missing Values"` and assigns it to the variable `fig_missing`.
- The seventh line displays the plot using the `show` method on the `fig_missing` variable.
- The eighth line returns the `missingInfo` dataframe as the output of the function.
- The ninth line calls the function with a dataframe as an argument and displays its output.

Result : The result shows the data types, the number of missing values, and the percentage of missing values for each column in the input dataframe. The columns are sorted by the percentage of missing values in descending order. The result indicates that the `Cabin` column has the most missing values (77.10%), followed by the `Age` column (19.87%), and the `Embarked` column (0.22%). The rest of the columns have no missing values. The result also shows that the columns have different data types, such as object, float64, and int64.

In [5]:
def missing_values(df):
    missingInfo = pd.DataFrame(df.dtypes, columns=["dtypes"])
    missingInfo["missing"] = df.isnull().sum()
    missingInfo["% missing"] = round((df.isnull().sum() / len(df)) * 100, 2)
    missingInfo = missingInfo.sort_values(by="% missing", ascending=False)
    fig_missing = px.imshow(df.isnull(), template="plotly_dark", width=500, title="Missing Values")
    fig_missing.show()
    return missingInfo
missing_values(df)

dtypes  missing  % missing
Cabin         object      687      77.10
Age          float64      177      19.87
Embarked      object        2       0.22
PassengerId    int64        0       0.00
Survived       int64        0       0.00
Pclass         int64        0       0.00
Name          object        0       0.00
Sex           object        0       0.00
SibSp          int64        0       0.00
Parch          int64        0       0.00
Ticket        object        0       0.00
Fare         float64        0       0.00
FamilySize     int64        0       0.00

In [6]:
# fill the missing values in the Age column with the mean of the column because it is a numerical column
df["Age"] = df["Age"].fillna(df["Age"].mean())
# fill the missing values in the Embarked column with the most frequent value of the column because it is a categorical column
df["Embarked"] = df["Embarked"].fillna(df["Embarked"].mode()[0])
# fill the missing values in the Cabin column with the string "Unknown" instead of dropping the column because there are too many missing values
df["Cabin"] = df["Cabin"].fillna("Unknown")
# call the missing_values function defined earlier to see the updated information about the missing values
missing_values(df)

dtypes  missing  % missing
PassengerId    int64        0        0.0
Survived       int64        0        0.0
Pclass         int64        0        0.0
Name          object        0        0.0
Sex           object        0        0.0
Age          float64        0        0.0
SibSp          int64        0        0.0
Parch          int64        0        0.0
Ticket        object        0        0.0
Fare         float64        0        0.0
Cabin         object        0        0.0
Embarked      object        0        0.0
FamilySize     int64        0        0.0

---

Objective : Analyze the distribution of data in a dataframe<br>
About the code : Create a figure with multiple subplots that show the distribution of data from a dataframe called `df`. The dataframe contains information about the passengers on the Titanic, such as their gender, class, embarked port, survival status, age and family size. The code uses the `plotly` library to create interactive plots that can be displayed in a web browser.

Justification : The code uses bar plots and pie plots for the categorical variables (gender, class, embarked port and survival status) because they can show the frequency and proportion of each category easily. The code chooses to use histograms for the numerical variables (age and family size) because they can show the shape and density of the data distribution.

The code follows these steps to build the chart:

- Define a function called `plot_row` that takes three arguments: `column`, `row` and `pull`. The function is used to create a bar plot and a pie plot for a given column of the dataframe in a specified row of the figure. The `pull` argument is used to pull out some slices of the pie plot for emphasis.
- Inside the function, use the `add_trace` method of the figure object to add a bar plot trace to the first column of the row. The x-axis values are the unique values of the column, the y-axis values are the counts of each value, and the text labels are also the counts. The name and marker color are left empty.
- Use the same method to add a pie plot trace to the second column of the row. The labels are the unique values of the column, the values are the counts of each value, and the name is left empty. The pull argument is passed to the pie plot trace to pull out some slices. The marker colors are chosen from a qualitative palette from `plotly.express`.
- Create the subplots layout using the `make_subplots` function. The function takes several arguments, such as:
  - `rows` and `cols`, which specify how many rows and columns of subplots are in the figure.
  - `specs`, which is a list of lists that defines what type of plot is in each subplot. In this case, each row has a bar plot and a pie plot, except for rows 5 and 7, which have only one histogram plot that spans two columns.
  - `subplot_titles`, which is a list of strings that gives titles to each subplot.
- Plot the rows using the `plot_row` function defined earlier. Pass in the column name, row number and pull argument for each row. For example, to plot gender bar and pie plots in row 1, use `plot_row('Sex', 1, pull=[0.1, 0])`. This means that pull out the first slice (male) by 0.1 units from the center of the pie.
- Plot the age histogram using the `add_trace` method. Pass in a histogram trace with x-axis values as
the age column of the dataframe, name as empty string, histnorm as density (to normalize
the histogram), and marker color as a hex code.
- Plot the family size histogram using the same method. Pass in a histogram trace with x-axis values as the family size column of the dataframe, name as empty string, histnorm as density, text labels as counts of each value, and marker colors as a diverging palette from `plotly.express`.
- Update the layout and traces using the `update_layout` and `update_traces` methods. Some of the layout options are:
  - `height` and `width`, which set the height and width of the figure in pixels.
  - `showlegend`, which controls whether to show legends or not.
  - `title_text`, which sets the title text of the figure.
  - `title_x`, which sets the position of the title along the x-axis.
  - `titlefont`, which sets the font style of the title.
  - `paper_bgcolor`, which sets the background color of the plot area.
  - `plot_bgcolor`, which sets the background color of the entire plot.
  - `font_color`, which sets the font color.
- Update some specific traces using filters or selectors. For example, use `update_yaxes(showgrid=False)` to remove y-axis gridlines from all subplots. Use `update_traces(marker_line_color='black', marker_line_width=2)` to change marker line color and width for only histogram traces. Use `update_annotations(font={'color': '#6bddff'})` to change or restyle title color for all subplots.
- Display the figure using the `show` method. This will open a web browser and show the interactive figure.

Results :

For gender, there were 577 male (35.2%) and 314 female (64.8%).

For passenger class, 216 people (24.2%) were in the upper class, 184 people (20.7%) were in the middle class, and 491 people (55.1%) were in the lower class.

In terms of where people boarded the ship, 646 people (72.5%) boarded in Southampton, 168 people (18.9%) boarded in Cherbourg, and 77 people (8.64%) boarded in Queenstown.
- Out of all the passengers, 342 (38.4%) survived, while 549 (61.6%) did not survive.

Now let's talk about the ages of the passengers. The average age of the people on the ship is about 29.54 years. This means that most of the people are around 29 years old.
- The middle age (median) is 29, which means that half of the people on the ship are younger than 29 and half are older than 29.
- The most common age (mode) is 29, which means that there are more people who are 29 years old than any other age.
- The range of ages is 80, which means that the oldest person on the ship is 80 years older than the youngest person.
- The interquartile range (IQR) of ages is 13, which shows that the middle 50% of the ages fall within a range of 13 years. This tells us how spread out the ages are.
- The standard deviation of ages is 13.01. This tells us that the ages vary quite a bit from the average age of 29. A higher standard deviation means more variability in the ages.

Now let's look at the family sizes. On average, each family has about 1.9 members.
- The median family size is 1.0, which means that half of the families have only one member, while the other half has more than one member.
- The most common family size is 1 member, which means that there are more families with just one person.
- The range of family sizes is 10, which means that the largest family has 10 more members than the smallest family.
- The interquartile range (IQR) of family sizes is 1.0, which shows that the middle 50% of families have a family size range of 1 member.
- The standard deviation of family sizes is approximately 1.6, which means that the family sizes vary quite a bit from the average value of 1.9.

In [7]:
def plot_row(column, row, pull=None):
    fig.add_trace(go.Bar(
        x=df[column].unique(),
        y=df[column].value_counts().values,
        text=df[column].value_counts(),
        name="",
        marker_color=None,
    ), row=row, col=1)

    fig.add_trace(go.Pie(
        labels=df[column].unique(),
        values=df[column].value_counts().values,
        name="",
        pull=pull,
        marker_colors=px.colors.qualitative.G10
    ), row=row, col=2)

fig = make_subplots(
    rows=8, cols=2,
    specs=[[{}, {'type': 'domain'}],
           [{}, {'type': 'domain'}],
           [{}, {'type': 'domain'}],
           [{}, {'type': 'domain'}],
           [{"rowspan": 2, "colspan": 2}, None],
           [None, None],
           [{"rowspan": 2, "colspan": 2}, None],
           [None, None],
          ],
    subplot_titles=('Gender Bar', 'Gender Pie', 'Pclass Bar', 'Pclass Pie', 'Embarked Bar', 'Embarked Pie',
                    'Survived Bar', 'Survived Pie', 'Age Distribution', 'Family Size'),
)

plot_row('Sex', 1, pull=[0.1, 0])
plot_row('Pclass', 2, pull=[0.1, 0, 0])
plot_row('Embarked', 3, pull=[0.1, 0, 0])
plot_row('Survived', 4, pull=[0.1, 0])

fig.add_trace(go.Histogram(
    x=df['Age'],
    name="",
    histnorm='density',
    marker_color="#f4a582"
), row=5, col=1)

fig.add_trace(go.Histogram(
    x=df['FamilySize'],
    name="",
    histnorm='density',
    text=df['FamilySize'].value_counts(),
    marker_color=px.colors.diverging.RdBu
), row=7, col=1)

fig.update_layout(
    height=1600,
    width=800,
    showlegend=False,
    title_text="Distribution of data",
    title_x=0.5,
    titlefont={'size': 25, 'family': 'Roboto', 'color': 'white'},
    paper_bgcolor="black",
    plot_bgcolor="black",
    font_color="white"
)

fig.update_yaxes(showgrid=False)

fig.update_traces(marker_line_color='black', marker_line_width=2)

fig.update_annotations(font={'color': '#6bddff'})

fig.show()

In [8]:
# Replace `data` with your actual age data
def histogram_summary(column):
    # Print the column name
    print("Summary for", column)
    
    # Extract the values from the specified column in the DataFrame
    data = df[column].values
    
    # Remove any NaN (missing) values from the data and convert to integers
    data = data[~np.isnan(data)].astype(int)

    # Calculate the mean, median, and mode of the data
    mean = np.mean(data)
    median = np.median(data)
    mode = np.argmax(np.bincount(data))

    # Calculate the range, interquartile range, and standard deviation of the data
    range1 = np.ptp(data)
    iqr = np.percentile(data, 75) - np.percentile(data, 25)
    std = np.std(data)

    # Print the calculated results
    print("Mean :", mean)
    print("Median:", median)
    print("Mode:", mode)
    print("Range:", range1)
    print("Interquartile range:", iqr)
    print("Standard deviation:", std, "\n")

# Call the histogram_summary function for the "Age" column
histogram_summary("Age")

# Call the histogram_summary function for the "FamilySize" column
histogram_summary("FamilySize")


Summary for Age
Mean : 29.544332210998878
Median: 29.0
Mode: 29
Range: 80
Interquartile range: 13.0
Standard deviation: 13.006473346327034 

Summary for FamilySize
Mean : 1.904601571268238
Median: 1.0
Mode: 1
Range: 10
Interquartile range: 1.0
Standard deviation: 1.6125528671095162 



---

Objective : To identify outliers in the Titanic dataset for certain variables. <br>
About the code : Creates a box plot visualization to identify outliers in the Titanic dataset. <br>
Justification : The chart was chosen because box plots are effective in visualizing the distribution and identifying outliers in numerical data.

Step-by-step explanation of the code:

1. Box traces are defined as a list of box plot objects using the `go.Box` function. Each box plot represents a different variable from the Titanic dataset: 'Fare', 'Age', 'SibSp', and 'Parch'. The color of each box plot is specified using the `marker` argument.

2. The layout of the chart is defined using the `go.Layout` function. The layout includes a title, a fixed height of 400 pixels, and a font size of 14. The template is set to "plotly_dark" for a dark-themed visualization.

3. A `go.Figure` object is created, which combines the box traces and the layout.

4. Finally, the `fig.show()` method is called to display the chart.

Results :

Outliers are special data points that are very different from the others. In the Titanic dataset, there are some passengers who have unusual information compared to the rest.

For 'Parch' (which stands for parents and children), 213 passengers have a higher number than 0. This means they were traveling with more family members than most people.

For 'SibSp' (which stands for siblings and spouses), 46 passengers have a higher number than 2.50. This means they had more siblings or spouses with them compared to most people.

For 'Age', 24 passengers have an age lower than 2.50. This is quite unusual because most passengers were older than that. It could be a mistake or a special case.

For 'Fare', 116 passengers paid a higher price than 65.63. This means they spent more money on their ticket compared to most people.

In [9]:
box_traces = [
    go.Box(x=df['Fare'], name='Fare', marker=dict(color='#87edff')),
    go.Box(x=df['Age'], name='Age', marker=dict(color='#f37a7a')),
    go.Box(x=df['SibSp'], name='SibSp', marker=dict(color='green')),
    go.Box(x=df['Parch'], name='Parch', marker=dict(color='orange'))
]

layout = go.Layout(
    title='Outliers in Titanic Dataset',
    height=400,
    font=dict(size=14),
    template = "plotly_dark"
)

fig = go.Figure(data=box_traces, layout=layout)

fig.show()

In [10]:
# Calculate summary statistics
fare_stats = df['Fare'].describe()
age_stats = df['Age'].describe()
sibsp_stats = df['SibSp'].describe()
parch_stats = df['Parch'].describe()

# Define the range for outliers (e.g., 1.5 times the interquartile range)
fare_outlier_range = 1.5 * (fare_stats['75%'] - fare_stats['25%'])
age_outlier_range = 1.5 * (age_stats['75%'] - age_stats['25%'])
sibsp_outlier_range = 1.5 * (sibsp_stats['75%'] - sibsp_stats['25%'])
parch_outlier_range = 1.5 * (parch_stats['75%'] - parch_stats['25%'])

# Identify outliers
fare_outliers = df[df['Fare'] > fare_stats['75%'] + fare_outlier_range]
age_outliers = df[df['Age'] < age_stats['25%'] - age_outlier_range]
sibsp_outliers = df[df['SibSp'] > sibsp_stats['75%'] + sibsp_outlier_range]
parch_outliers = df[df['Parch'] > parch_stats['75%'] + parch_outlier_range]

# Print outlier information
print("Outlier Information:")
print(f"Parch Outliers (Above {parch_stats['75%'] + parch_outlier_range:.2f}): {len(parch_outliers)} passengers")
print(f"SibSp Outliers (Above {sibsp_stats['75%'] + sibsp_outlier_range:.2f}): {len(sibsp_outliers)} passengers")
print(f"Age Outliers (Below {age_stats['25%'] - age_outlier_range:.2f}): {len(age_outliers)} passengers")
print(f"Fare Outliers (Above {fare_stats['75%'] + fare_outlier_range:.2f}): {len(fare_outliers)} passengers")


Outlier Information:
Parch Outliers (Above 0.00): 213 passengers
SibSp Outliers (Above 2.50): 46 passengers
Age Outliers (Below 2.50): 24 passengers
Fare Outliers (Above 65.63): 116 passengers


---

Objective : To analyze the correlation between the features in the Titanic dataset. <br>
About the code : Creates a correlation matrix chart to visualize the pairwise correlations among the features in a DataFrame. <br>
Justification : The chart is chosen because it provides a clear and concise representation of the correlations.

Step-by-step explanation of the code:

1. Calculate the correlation matrix using the Pearson correlation method:
   <br>`corr = df.corr(method='pearson')`

2. Create a boolean mask with `True` values in the upper triangular part of the correlation matrix:
   <br>`mask = np.triu(np.ones_like(corr, dtype=bool))`

3. Apply the mask to the correlation matrix, replacing the values in the upper triangular part with NaN (masked values):
   <br>`masked_corr = corr.mask(mask)`

4. Create a new Plotly figure using `px.imshow()` and specify the title, height, width, and template:
   <br>`fig = px.imshow(masked_corr, title='Correlations Among Features', height=700, width=700, template="plotly_dark")`

5. Set the text for each cell in the plot to the corresponding correlation value rounded to 2 decimal places:
   <br>`fig.update_traces(text=corr.values.round(2), hovertemplate='Feature 1: %{y}<br>Feature 2: %{x}<br>Correlation: %{text}')`

6. Add a colorbar to the chart with a title:
   <br>`fig.update_traces(colorbar=dict(title="Correlation"))`

7. Iterate over each cell in the masked correlation matrix:
   - If the cell is not on the diagonal (i != j) and it has a non-null value, do the following:
     - Determine the color for the annotation text based on the correlation value (white if < 0.5, black otherwise).
     - Add an annotation with the correlation value to the corresponding cell in the figure:
       - `fig.add_annotation(x=j, y=i, text=str(masked_corr.iloc[i, j].round(2)), showarrow=False, font=dict(color=color))` 
<br><br>
8. Update the x-axis to remove gridlines:
   <br>`fig.update_xaxes(showgrid=False)`

9. Update the y-axis to remove gridlines:
   <br>`fig.update_yaxes(showgrid=False)`

10. Display the final chart:
    <br>`fig.show()`

Results :
- highest positive correlation: 'SibSp' and 'Parch' (excluded FamilySize)
- highest negative correlation: 'Pclass' and 'Fare'
- highest positive correlation with Survived: 'Fare'
- highest negative correlation with Survived: 'Pclass'


In [11]:
corr = df.corr(method='pearson')

mask = np.triu(np.ones_like(corr, dtype=bool))

masked_corr = corr.mask(mask)

fig = px.imshow(masked_corr,
                title='Correlations Among Features',
                height=700, width=700, template="plotly_dark")

fig.update_traces(text=corr.values.round(2),
                  hovertemplate='Feature 1: %{y}<br>Feature 2: %{x}<br>Correlation: %{text}',
                  colorbar=dict(title="Correlation"))

for i in range(len(masked_corr)):
    for j in range(len(masked_corr)):
        if i != j and not pd.isnull(masked_corr.iloc[i, j]):
            color = 'white' if float(masked_corr.iloc[i, j]) < 0.5 else 'black'
            fig.add_annotation(x=j, y=i, text=str(masked_corr.iloc[i, j].round(2)),
                               showarrow=False, font=dict(color=color))

fig.update_xaxes(showgrid=False)
fig.update_yaxes(showgrid=False)

fig.show()

---

Objective : To analyze the relationship between Age, Fare, Pclass, and Survival <br>
About the code : Creates a 3D scatter plot using Plotly, visualizing the relationship between age, fare, passenger class (Pclass), and survival (Survived) in a dataset (df). <br> 
Justification : The chart was chosen to visualize the relationship between age, fare, passenger class, and survival in a dataset because a 3D scatter plot allows for the simultaneous representation of three numerical variables (age, fare, and passenger class) as well as a categorical variable (survival). This type of chart can provide insights into potential patterns or correlations between these variables.

Here is a step-by-step explanation of each line and part of the code:

1. `color_sequence = ['#8B0000', '#00FF00']`: This line defines a list of two color codes, `'#8B0000'` (dark red) and `'#00FF00'` (green), which will be used to represent the different categories of the "Survived" variable.

2. `fig = px.scatter_3d(df, x="Age", y="Fare", z="Pclass", color="Survived", color_continuous_scale=color_sequence, title="Age vs Fare vs Class by Survived", template="plotly_dark", height=700, width=700)`: This line creates a 3D scatter plot using the Plotly Express (px) library. It specifies the dataset `df` and assigns the variables "Age", "Fare", and "Pclass" to the x, y, and z axes, respectively. The "Survived" variable is used to determine the color of each data point, with the colors defined by `color_sequence`. Additional parameters include the chart title ("Age vs Fare vs Class by Survived"), the template ("plotly_dark" for a dark-themed plot), and the dimensions of the chart (height=700, width=700).

3. `fig.show()`: This line displays the created figure (`fig`) in the output.


Results :

Green dots represent passengers who survived, while red dots represent passengers who did not survive.
Passengers in higher passenger classes had higher survival rates compared to lower classes. Those who paid higher fares also had higher chances of survival. Additionally, younger passengers had higher survival rates compared to older passengers. In summary, higher class, higher fare, and younger age were associated with increased chances of survival.

In [12]:
color_sequence = ['#8B0000', '#00FF00']
fig = px.scatter_3d(df, x="Age", y="Fare", z="Pclass", color="Survived", color_continuous_scale=color_sequence, title="Relationship between Age, Fare, Pclass, and Survival", template="plotly_dark", height=700, width=700)
fig.show()